guides/batching.md

```diff
@@ -1,7 +1,201 @@
+  end
+end
+```
+
+Now we can use this function to batch our author lookups:
+
+```elixir
+object :post do
+
+  @desc "Author of the post"
+  field :author, :user do
+    resolve fn post, _, _ ->
+      batch({MyAppWeb.Schema.Helpers, :by_id, User}, post.author_id, fn batch_results ->
+        {:ok, Map.get(batch_results, post.author_id)}
+      end)
+    end
+  end
+
+end
+```
+
+Now we make just two calls to the database. The first call loads all of the posts.
+Then as Absinthe walks through each post and tries to get the author, it's instead
+told to aggregate its information. That aggregate information is passed on to our `by_id/2` function from earlier.
+It grabs ALL the users in just one database call, and creates a map of user ids
+to users.
+
+Absinthe then does a second pass and calls the `batch_results` function with that
+map, letting us retrieve the individual author for each post.
+
+Not only is this a very efficient way to query the data, it's also 100% dynamic.
+If a query document asks for authors, they're loaded efficiently. If it does not,
+they aren't loaded at all.
 
 See the documentation for [Absinthe.Middleware.Batch](Absinthe.Middleware.Batch.html) for more information.
+
+`Absinthe.Middleware.Batch` achieves a lot and, with some helpers, was the
+standard way to solve this problem for a long time. While batching still has a
+place, it has a few limitations that have driven the development of Dataloader.
+There are small scale annoyances like the limitation of only being able to batch
+one thing at a time in a field, or the fact that the API can get very verbose.
+
+There's also some larger scale issues however. Ecto has a fair number of quirks
+that make it a difficult library to abstract access to. If you want the
+concurrent test system to work, you need to add `self()` to all the batch keys
+and do `Repo.all(caller: pid)` in every batch function so that it knows which
+sandbox to use. It gets very easy for your GraphQL functions to become full of
+direct database access, inevitably going around important data access rules you
+may want to enforce in your contexts. Alternatively, your context functions can
+end up with dozens of little functions that only exist to support batching items
+by ID.
+
+In time, people involved in larger projects have been able to build some
+abstractions, helpers, and conventions around the `Absinthe.Middleware.Batch`
+plugin that have done a good job of addressing these issues. That effort has been
+extracted into the Dataloader project, which also draws inspiration from similar
+projects in the GraphQL world.
+
+We've made it easier and more flexible, however, with
+Elixir's [dataloader](https://hex.pm/packages/dataloader) package.
+
+### Dataloader
+
+Let's jump straight in to getting Dataloader working, and then we'll expand on
+what's actually happening behind the scenes.
+
+Using Dataloader is as simple as doing:
+
+```elixir
+import Absinthe.Resolution.Helpers, only: [dataloader: 1]
```

Comment:
(Being slightly nit-picky) We might want something like the following here:

```elixir
alias MyApp.Blog # Dataloader source, see below
```

The concept of a source is still being introduced. I think we should call it out explicitly because this code block might be confusing to someone who doesn't already know what `Blog` is supposed to be an example of.

Reply:
Fixed by: 3732ad8